Capture hw mute indicator

jsarha opened the pull request:
There are some missing pieces in this code, but I wanted to make it available before I go to short vacation in the end of the week. The missing pieces are:
- no support for handling SDW linux driver hw mute status updates
- the snd_ipc4_global_capture_hw_mute_force() callback is not connected to DSP power-state (have not found an appropriate place yet... reverse engineering continues)

To test the code, this SOF  draft PR can be used: https://github.com/thesofproject/sof/pull/9800
When integrating the pieces to SOF mic_privacy code, the driver side expects a privacy state update when the DSP has booted up.

There are still unsolved issues to this concept (how to combine SDW/host-linux and DMIC/sof-FW mute status to one indicator) and they need to be resolved, before this work can be finished.

Commit:
ASoC: sof: ipc4-control: Global mixers and capture hw mute indicator

Implement global kcontrol concept and add capture hw mute indicator
implementation in it.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: SOF: ipc4: Add enum for global kcontrol IDs

Add enum for global kcontrol IDs and
SOF_IPC4_KCONTROL_GLOBAL_CAPTURE_HW_MUTE in it.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: SOF: IPC4: Add SOF_IPC4_FW_GLOBAL_KCONTROL_MASK to fw_config_params

Add SOF_IPC4_FW_GLOBAL_KCONTROL_MASK to fw_config_params. The purpose
is to inform the driver which global kcontrols are supported by the
FW.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: sof: ipc4-control: Force method for global capture hw mute indicator

This is inteded for driving the mute indicator following dsp power
state. If dsp is powerred down the capture muted state should be
indicated.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -690,6 +695,14 @@ static void sof_ipc4_control_update(struct snd_sof_dev *sdev, void *ipc_message)
+	if (ndata->module_id == SOF_IPC4_MOD_INIT_BASEFW_MOD_ID &&
+	    ndata->instance_id == SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID) {
+		dev_info(sdev->dev, "Received global notifier event_id %#x",
+			event_param_id);
```
this should be dev_dbg()

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -840,6 +852,140 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+	int ret;
+
+	if (!snd_card) {
+		dev_err(dev, "%s: Card not found!", __func__);
```
you have multiple prints in sof_scomp_get_card()

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -840,6 +852,140 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+	if (!scomp->card) {
+		dev_err(dev, "%s: No snd_soc_card", __func__);
+		return NULL;
+	}
```
is any of these can happen?

jsarha replied:
Forgot them there. Its just that I did not exactly know what I was doing, and I did not want to start debugging crashes, I  checked every possible problem in the first version

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -840,6 +852,140 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+		return -ENODEV;
+	}
+
+	if (!ipc4_data) {
```
I don't think this can happen if you are here.

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -840,6 +852,140 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+	struct device *dev = sdev->dev;
+
+	dev_info(dev, "%s: got msg id %u num_elems %u", __func__,
+		 msg_data->id, msg_data->num_elems);
```
dev_dbg()

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -649,6 +649,10 @@ sof_ipc4_volsw_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
 
+
+static void snd_ipc4_handle_global_event(struct snd_sof_dev *sdev,
+					 struct sof_ipc4_control_msg_payload *msg_data);
```
why don't you add the function to a place where you don't need forward declaration? That would be cleaner

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -840,6 +852,140 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+			snd_ipc4_global_capture_hw_mute_report(sdev, msg_data->chanv[0].value);
+		else
+			dev_warn(dev, "%s: bad data for id %u chan 0 %u", __func__,
+				 msg_data->id, msg_data->chanv[0].channel);
```
is it the channel or the num_elems are the 'bad' ones? if it is the num_elems, we will have to guess what it is.

ujfalusi replied:
num_elems = 42 is fine here if chanv[0].channel is 0?

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -840,6 +852,140 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+
+	kctl = priv->capture_hw_mute;
+
+	dev_dbg(dev, "%s: reporting %u, old %lu", __func__,
```
dev_dbg() does not need `__func__` to be printed.

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -840,6 +852,140 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+
+		ret = snd_ctl_add(snd_card, kctl);
+		if (ret < 0)
+			dev_err(dev, "snd_ctl_add(): failed");
```
for these errors I would add `__func__`

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -949,13 +956,19 @@ static void snd_ipc4_global_capture_hw_mute_report(struct snd_sof_dev *sdev,
-	kctl = priv->capture_hw_mute;
+	kctl = priv->capture_hw_mute.kctl;
+
+	dev_dbg(dev, "%s: new %u, old %lu, last %u force %u", __func__,
```
No `__func__`

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -854,7 +854,11 @@ sof_ipc4_set_up_volume_table(struct snd_sof_control *scontrol, int tlv[SOF_TLV_I
+	struct sof_ipc4_kcontrol_global_capture_hw_mute {
+		struct snd_kcontrol *kctl;
+		bool force_muted; /* If true mute state indicator forced to muted */
+		bool last_reported; /* Last statusmute state received from FW */
```
Comment needs a bit of work ;)

ujfalusi commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -964,6 +977,16 @@ static void snd_ipc4_global_capture_hw_mute_report(struct snd_sof_dev *sdev,
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_ipc4_kcontrol_global_priv *priv = ipc4_data->global_kcontrol_priv;
+
+	priv->capture_hw_mute.force_muted = force;
```
Here you need to check for NULL pointers

ujfalusi commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -119,4 +119,6 @@ void sof_ipc4_update_cpc_from_manifest(struct snd_sof_dev *sdev,
 size_t sof_ipc4_find_debug_slot_offset_by_type(struct snd_sof_dev *sdev,
 					       u32 slot_type);
 
+void snd_ipc4_global_capture_hw_mute_force(struct snd_sof_dev *sdev, bool force);
```
I guess the calling of this is under works?

jsarha replied:
Any good suggestion of where to follow DSP powerstate is apreciated.

Commit:
ASoC: SOF: Intel: Force global capture hw mute according to DSP power

Force global capture hw mute to muted state when DSP is powered
down. This is done by calling snd_ipc4_global_capture_hw_mute_force()
from hda_dsp_set_power_state_ipc4().

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

jsarha commented:
Now the PR contains also the code force the capture hw mute indicator to muted state, when DSP is powered down.
